pandas/core/reshape/pivot.py

```diff
@@ -424,15 +425,19 @@ def _convert_by(by):
 
 @Substitution("\ndata : DataFrame")
 @Appender(_shared_docs["pivot"], indents=1)
-def pivot(data: "DataFrame", index=None, columns=None, values=None) -> "DataFrame":
+def pivot(
+    data: "DataFrame",
+    index: Optional[Union[Label, Sequence[Optional[Label]]]] = None,
+    columns: Optional[Union[Label, Sequence[Optional[Label]]]] = None,
+    values: Optional[Union[Label, Sequence[Optional[Label]]]] = None,
+) -> "DataFrame":
     if columns is None:
         raise TypeError("pivot() missing 1 required argument: 'columns'")
     columns = columns if is_list_like(columns) else [columns]
 
     if values is None:
-        cols: List[str] = []
         if index is None:
-            pass
+            cols = []
         elif is_list_like(index):
             cols = list(index)
```

Comment:
One of the mypy failures is that `is_list_like` should narrow the scope of what goes down this branch, but mypy doesn't recognize it and thinks you could potentially be calling `list(index)` with a `Hashable`. That yields `error: No overload variant of "list" matches argument type "Hashable"`

@jbrockmendel or @simonjayhawkins I forget, but do we have a preferred way of handling that scoping? Would it be `assert isinstance(index, abc.Sequence)` here?

Reply:
is this something we expect mypy to get smarter about?

im increasingly leaning towards "type: ignore" on the grounds that it a) has zero runtime overhead and b) is easy to grep for in periodic see-if-we-can-remove-this sweeps

Reply:
Maybe but its not implemented. The mypy issue number is 5206 so I think OK to also do `#type: ignore - mypy 5206` or something similar

Reply:
I think once we have #29197 in play, the #type: ignore could be a better option. so we could start adding type: ignore and once we have #29197, go back and ensure that the ignores are correct.

Reply:
> Would it be `assert isinstance(index, abc.Sequence)` here

I _think_ asserts are OK when narrowing to exclude a type, e.g. assert is not None, but cast is better for narrowing to ensure a type. ( unless the assert is beneficial in a run-time context)

Reply:
@WillAyd @jbrockmendel #32785 was opened for this discussion. the #type: ignore option could be discussed there also. cc @jorisvandenbossche